msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java

```diff
@@ -1263,6 +1267,130 @@ public boolean isSharedDevice() {
         return mPublicClientConfiguration.getIsSharedDevice();
     }
 
+    @Override
+    public String generateSignedHttpRequest(@NonNull final IAccount account,
+                                            @NonNull final PoPAuthenticationScheme popParameters) throws MsalException {
+        final ResultFuture<AsyncResult<GenerateShrResult>> future = new ResultFuture<>();
+
+        final GenerateShrCommandParameters cmdParams =
+                CommandParametersAdapter.createGenerateShrCommandParameters(
+                        mPublicClientConfiguration,
+                        mPublicClientConfiguration.getOAuth2TokenCache(),
+                        ((Account) account).getHomeAccountId(),
+                        popParameters
+                );
+        final GenerateShrCommand generateShrCommand = new GenerateShrCommand(
+                cmdParams,
+                MSALControllerFactory.getAllControllers(
+                        mPublicClientConfiguration.getAppContext(),
+                        mPublicClientConfiguration.getDefaultAuthority(),
+                        mPublicClientConfiguration
+                ),
+                new CommandCallback<GenerateShrResult, BaseException>() {
+                    @Override
+                    public void onCancel() {
+                        // Not cancellable
+                    }
+
+                    @Override
+                    public void onError(@NonNull final BaseException error) {
+                        future.setResult(
+                                new AsyncResult<GenerateShrResult>(
+                                        null,
+                                        baseExceptionToMsalException(error)
+                                )
+                        );
+                    }
+
+                    @Override
+                    public void onTaskCompleted(@NonNull final GenerateShrResult generateShrResult) {
+                        future.setResult(new AsyncResult<>(generateShrResult, null));
+                    }
+                },
+                "1100" // Public API ID
+        );
+
+        // Execute this command silently...
+        CommandDispatcher.submitSilent(generateShrCommand);
+
+        try {
+            final AsyncResult<GenerateShrResult> asyncResult = future.get();
+
+            if (asyncResult.getSuccess()) {
+                return asyncResult.getResult().getShr();
+            } else {
+                throw asyncResult.getException();
+            }
+        } catch (ExecutionException | InterruptedException e) {
+            throw new MsalClientException(
+                    UNKNOWN_ERROR,
+                    "Unexpected error while generating SHR.",
+                    e
+            );
+        }
+    }
```

Comment:
Instead of writing all this, could potentially just call the overload below by passing a custom callback that sets the result on the future....so we can avoid this mostly identical code.

Reply:
2d3ded5873f0e1164a390059887b1de952e5d055